Episode 3 - connecting to MUSCLE

JaroCamphuijsen opened the pull request:
Episode 3 is ready. The MUSCLE part is split up into episode 3 and 4. This first part is about how to connect a submodel to MUSCLE. Connecting and running the coupled simulation will come in episode 4.

JaroCamphuijsen committed:
split muscle episode into connect and couple

JaroCamphuijsen committed:
add introduction and first challenge to episode 3

JaroCamphuijsen committed:
improve challenge description

JaroCamphuijsen committed:
add exercises and all unpolished content for episode 3

JaroCamphuijsen committed:
fine tune episode 3, add extra challenge

JaroCamphuijsen commented on episodes/03-muscle-connect.md:
````diff
@@ -0,0 +1,366 @@
+
+```python
+    # begin state_update_loop
+        while t_cur + dt < msg.timestamp + t_max:
````
@LourensVeen Isn't it more efficient to leave the state update loop as it was: `while t_cur + dt < t_max`
and make sure that the received timestamp is added to t_cur when a global time stamp is sent out? Is there a good reason to do this addition every time in the while loop?

LourensVeen replied:
Yes, `msg.timestamp` and `t_max` don't change while we're iterating, so we could do that addition once, after receiving `msg` and before starting the loop.

We cannot add the received timestamp to `t_cur` though, as both are absolute times. Every time we enter the state update loop, we need to run from `msg.timestamp` to `msg.timestamp + t_max`. Does that make sense?

JaroCamphuijsen replied:
I changed it in my latest commit to add the timestamp in the initialisation already, the reuse loop then stays as simple as possible and unchanged from the code we start with. See this commit: https://github.com/esciencecenter-digital-skills/model-coupling-workshop/pull/2/commits/eb0af7e9d0ec6b829b6f585203708951b37b8f5a

LourensVeen replied:
Okay, yes this looks good.

The pedantic hairsplitter in the back of my mind may want to change t_max to t_end at some point, but it would probably be good to keep this and the MUSCLE3 docs in sync, and it can wait. So let's keep it at this, nice job!

JaroCamphuijsen committed:
change time tracking in example

JaroCamphuijsen committed:
adjust exercise numbering

raar1 commented on episodes/03-muscle-connect.md:
```diff
@@ -0,0 +1,418 @@
+
+**ymmsl-python** is a Python library that contains class definitions to represent an MMSL model description. A yMMSL file serves as the interface between us (humans) and the muscle manager and is meant to describe the multiscale simulation and tell the muscle manager what to do.
+
+We continue with the reaction-diffusion model example that we used the previous episode and in this episode we will connect an implementation of the reaction model to the MUSCLE3 library step by step.
```
```suggestion
We continue with the reaction-diffusion model example that we used in the previous episode and in this episode we will connect an implementation of the reaction model to the MUSCLE3 library step by step.
```

raar1 commented on episodes/03-muscle-connect.md:
```diff
@@ -0,0 +1,418 @@
+
+We continue with the reaction-diffusion model example that we used the previous episode and in this episode we will connect an implementation of the reaction model to the MUSCLE3 library step by step.
+
+## Disect your model
```
```suggestion
## Dissect your model
```

raar1 commented on episodes/03-muscle-connect.md:
```diff
@@ -0,0 +1,418 @@
+
+## Exercise 1: Can you recognize the Submodel Execution Loop?
+
+Open the file called reaction.py in a text editor. Can you recognize the beginning and end of the four operators ($F_{init}$, $O_I$, $S$ and $O_F$ ) plus the state update loop in this submodel? Mark them with comments in the code (e.g. `# begin F_init`, 10 comments in total).
```
```suggestion
Open the file called reaction.py in a text editor (see below for code). Can you recognize the beginning and end of the four operators ($F_{init}$, $O_I$, $S$ and $O_F$ ) plus the state update loop in this submodel? Mark these by placing the following 10 comments in the code:
- `# begin F_INIT`
- `# end F_INIT`
- `# begin O_I`
- `# end O_I`
- `# begin S`
- `# end S`
- `# begin state_update_loop`
- `# end state_update_loop`
- `# begin O_F`
- `# end O_F`
```

raar1 replied:
Feel free to reject this suggestion, but I think it's clearer in the exercise  description to explicitly list the 10 comments we want them to place in the code.

raar1 commented on episodes/03-muscle-connect.md:
````diff
@@ -0,0 +1,418 @@
+    # end O_F
+```
+
+The Submodel Execution Loop specifies two operators within the state update loop. $O_I$ and comes before $S$ and you can use it to send a message to the outside world with (part of) the current state. In this case, the $O_I$ operator is empty; we are not sending or storing any intermediate states. The original type annotations in the function definition serve the same purpose but we will be losin ...
````
`$O_I$ and comes before`
Remove the "and"? Or was there another operator here?

raar1 commented on episodes/03-muscle-connect.md:
```diff
@@ -0,0 +1,418 @@
+
+## The constructor
+
+To let a model communicate with the muscle manager, other submodels and the outside world we need to create a `libmuscle.Instance` object:
```
```suggestion
To let a model communicate with the muscle manager, other submodels and the outside world, we need to create a `libmuscle.Instance` object:
```

raar1 commented on episodes/03-muscle-connect.md:
````diff
@@ -0,0 +1,418 @@
+
+:::: solution
+
+```python
````
"Exercise 2: Constructing your submodel" suggests (to me) that you should have a working/finished submodel at the end of this exercise. I would therefore probably attempt to replace the `F_INIT` section with code that gets the values from `initial_state` and then get stuck. Is there a way we can clarify here that the "solution" will only be a work in progress, with further steps needed to finish the submodel?

JaroCamphuijsen replied:
I solved this by renaming the exercise, let me know if you think this is enough.

raar1 commented on episodes/03-muscle-connect.md:
```diff
@@ -0,0 +1,418 @@
+
+## Settings
+
+Next is the first part of the model, in which the model is initialised. Nearly every model needs some settings that define how it behaves (e.g. the size of a timestep or model specific parameters). We would like to control these settings centrally suing the MUSCLE manager when starting up the coupled simulation so we don't have to change it in our model code every time if we want to try a range o ...
```
```suggestion
Next is the first part of the model, in which the model is initialised. Nearly every model needs some settings that define how it behaves (e.g. the size of a timestep or model specific parameters). We would like to control these settings centrally using the MUSCLE manager when starting up the coupled simulation so we don't have to change it in our model code every time if we want to try a range of values (for instance, to perform a sensitivity analysis). We can use the `libmuscle.Instance.get_setting` function instead to ask the MUSCLE manager for the values. The values, or ranges of values, should be recorded in the configuration file which we will cover in the next chapter.
```

raar1 commented on episodes/03-muscle-connect.md:
```diff
@@ -0,0 +1,418 @@
+::: challenge
+
+## Exercise 3: 
+In our example several settings have been hard-coded into the model:
```
```suggestion
In our example, several settings have been hard-coded into the model:
```

raar1 commented on episodes/03-muscle-connect.md:
```diff
@@ -0,0 +1,418 @@
+## Exercise 3: 
+In our example several settings have been hard-coded into the model:
+
+- the total simulation time to run this sub-model `tmax`
```
```suggestion
- the total simulation time to run this sub-model, `tmax`
```

raar1 commented on episodes/03-muscle-connect.md:
```diff
@@ -0,0 +1,418 @@
+In our example several settings have been hard-coded into the model:
+
+- the total simulation time to run this sub-model `tmax`
+- the time step to use `dt`
```
```suggestion
- the time step to use, `dt`
```

raar1 commented on episodes/03-muscle-connect.md:
```diff
@@ -0,0 +1,418 @@
+
+- the total simulation time to run this sub-model `tmax`
+- the time step to use `dt`
+- and the model parameter `k`
```
```suggestion
- and the model parameter, `k`
```

raar1 commented on episodes/03-muscle-connect.md:
```diff
@@ -0,0 +1,418 @@
+## Receiving messages
+Apart from settings, we can use the `libmuscle.Instance.receive` function to receive an initial state for this submodel on the `initial_state` port. Note that we have declared that port above, and declared it to be associated with the `F_INIT` operator. During `F_INIT`, messages can only be received, not sent, so that declaration makes `initial_state` a receiving port.
+
+The message that we will receive can contain several pieces of information. For now, we are interested in the `data` and `timestamp` attributes. We assume the data to be a grid of floats containing our initial state and the time stamp tells us the simulated time at which this state is valid. We can store receive a message and store the `data` and `timestamp` attributes in the following way:
```
```suggestion
The message that we will receive can contain several pieces of information. For now, we are interested in the `data` and `timestamp` attributes. We assume the data to be a grid of floats containing our initial state and the time stamp tells us the simulated time at which this state is valid. We can receive a message and store the `data` and `timestamp` attributes in the following way:
```

?

raar1 commented on episodes/03-muscle-connect.md:
```diff
@@ -0,0 +1,418 @@
+
+## State update loop
+
+The actual state update happens in the operator $S$ in the Submodel Execution Loop, and in our example model, it is determined entirely by the current state. Since no information from outside is needed, we do not receive any messages, and in our `libmuscle.Instance` declaration above, we did not declare any ports associated with `Operator.S`. 
```
```suggestion
The actual state update happens in the operator $S$ in the Submodel Execution Loop and, in our example model, it is determined entirely by the current state. Since no information from outside is needed, we do not receive any messages, and in our `libmuscle.Instance` declaration above, we did not declare any ports associated with `Operator.S`. 
```

raar1 commented on episodes/03-muscle-connect.md:
```diff
@@ -0,0 +1,418 @@
+
+The operator `O_I` provides for observations of intermediate states. In other words, here is where you can send a message to the outside world with (part of) the current state. In this case, the `O_I` operator is empty; we’re not sending anything.
+
+One thing we need to be aware of is time. Most models depend on time in some way and it is good to be aware of the various time frames. MUSCLE3 enables passing on timestamps 
```
I'm not really sure what "MUSCLE3 enables passing on timestamps" means

JaroCamphuijsen replied:
Ah this was a line of thought that was lost at some point... I completed the sentence.

raar1 commented on episodes/03-muscle-connect.md:
```diff
@@ -0,0 +1,418 @@
+::: callout
+We convert our `numpy.array` explicitly into a `libmuscle.Grid object`, so that we can add the name of the dimensions. In our example case there is only one, and "x" is not very descriptive, so we could have also passed the array directly, in which case MUSCLE would have done the conversion for us and sent a `libmuscle.Grid` without index names automatically.
+
+Note that grids (and NumPy arrays) are much more efficient than lists and dictionaries. If you have a lot of data to send, then you should use those as much as possible. For example, a set of agents is best sent as a dictionary of 1D NumPy arrays, with one array/grid for each attribute. A list of dictionaries will be quite a bit slower and use a lot more memory, but it should work up to a million ...
```
Will all participants know what is meant here by a "set of agents"?

JaroCamphuijsen replied:
@raar1 Maybe we should remove this piece altogether. It doesn't really fit in this introduction tutorial to mention such a optimization detail.  Just remove line 341 altogether. Do you agree @LourensVeen ?
```suggestion
```

raar1 commented on episodes/03-muscle-connect.md:
```diff
@@ -0,0 +1,418 @@
+Note that grids (and NumPy arrays) are much more efficient than lists and dictionaries. If you have a lot of data to send, then you should use those as much as possible. For example, a set of agents is best sent as a dictionary of 1D NumPy arrays, with one array/grid for each attribute. A list of dictionaries will be quite a bit slower and use a lot more memory, but it should work up to a million ...
+:::
+
+The optional second parameter is a second timestamp, which is only used in [more advance applications](https://muscle3.readthedocs.io/en/latest/tutorial.html#message-timestamps) and can be set to `None` here.
```
```suggestion
The optional second parameter is a second timestamp, which is only used in [more advanced applications](https://muscle3.readthedocs.io/en/latest/tutorial.html#message-timestamps) and can be set to `None` here.
```

raar1 commented on episodes/03-muscle-connect.md:
````diff
@@ -0,0 +1,418 @@
+my_message = Message(my_timestamp, None, Grid(my_state, ['x']))
+```
+
+To send the message we specify the port on which to send which again needs to match the declaration.
````
```suggestion
To send the message we specify the port on which to send, which again needs to match the declaration.
```

raar1 commented on episodes/03-muscle-connect.md:
```diff
@@ -0,0 +1,418 @@
+::: keypoints
+
+- MUSCLE3 implements the MMSF theory, using the same terminology and MMSL for configuration of the coupled simulation
+- First analyze your code to    find the Submodel Execution Loop structure
```
```suggestion
- First analyze your code to find the Submodel Execution Loop structure
```

raar1 commented on episodes/03-muscle-connect.md:
```diff
@@ -0,0 +1,418 @@
+
+::: challenge
+
+## Exercise 1: Can you recognize the Submodel Execution Loop?
```
It could first be helpful to the learners if we introduce the (admittedly simple) reaction model - what is it actually modelling? (Or will this be in the earlier theory episode already?)

JaroCamphuijsen replied:
That depends a bit on Lourens' theory section. I suppose it would fit in there quite well since the model gets introduced there. But maybe a short description/walk through the code would fit here. I will add it.

raar1 reviewed (approved):
Great work @JaroCamphuijsen! I've left a couple of comments here and there, and some suggested text changes.

JaroCamphuijsen committed:
Update episodes/03-muscle-connect.md

Co-authored-by: Robin Richardson <raar1@users.noreply.github.com>

JaroCamphuijsen commented on episodes/03-muscle-connect.md:
````diff
@@ -0,0 +1,418 @@
+    # end O_F
+```
+
+The Submodel Execution Loop specifies two operators within the state update loop. $O_I$ and comes before $S$ and you can use it to send a message to the outside world with (part of) the current state. In this case, the $O_I$ operator is empty; we are not sending or storing any intermediate states. The original type annotations in the function definition serve the same purpose but we will be losin ...
````
```suggestion
The Submodel Execution Loop specifies two operators within the state update loop. $O_I$ comes before $S$ and you can use it to send a message to the outside world with (part of) the current state. In this case, the $O_I$ operator is empty; we are not sending or storing any intermediate states. The original type annotations in the function definition serve the same purpose but we will be losing these once we are done connecting the model to MUSCLE.
```

JaroCamphuijsen commented on episodes/03-muscle-connect.md:
```diff
@@ -0,0 +1,418 @@
+
+::: challenge
+
+## Exercise 2: Constructing your submodel
```
```suggestion
## Exercise 2: The MUSCLE constructor
```

JaroCamphuijsen commented on episodes/03-muscle-connect.md:
```diff
@@ -0,0 +1,418 @@
+
+The operator `O_I` provides for observations of intermediate states. In other words, here is where you can send a message to the outside world with (part of) the current state. In this case, the `O_I` operator is empty; we’re not sending anything.
+
+One thing we need to be aware of is time. Most models depend on time in some way and it is good to be aware of the various time frames. MUSCLE3 enables passing on timestamps 
```
```suggestion
One thing we need to be aware of is time. Most models depend on time in some way and it is good to be aware of the various time frames. Libmuscles messages support passing on timestamps to be able to keep track of the global simulation time of the coupled model. It is easier to keep track of only one time frame and it is therefore good practice to add the timestamp that accompanied the initial state to the sub-model simulation time steps instead of starting at zero for every sub-model instance. 
```

JaroCamphuijsen committed:
Apply suggestions from code review

Minor text edits and add explanation about timestamps and keeping track of time

Co-authored-by: Robin Richardson <raar1@users.noreply.github.com>

JaroCamphuijsen committed:
Fix broken list layout in episode 3

JaroCamphuijsen committed:
clarify the python code in episode 3, add titles to callouts

JaroCamphuijsen commented:
@raar1 in the github actions check it reports that there are errors in the lesson structure, but when I build it locally there is no problem. I'm not sure what I can do about that.

raar1 committed:
Try adding renv/sandbox to .gitignore

raar1 commented:
> @raar1 in the github actions check it reports that there are errors in the lesson structure, but when I build it locally there is no problem. I'm not sure what I can do about that.

@JaroCamphuijsen I was able to fix it by adding 'renv/sandbox' to the .gitignore

No idea why that causes the validation to choke, but nevermind...

JaroCamphuijsen commented:
> > @raar1 in the github actions check it reports that there are errors in the lesson structure, but when I build it locally there is no problem. I'm not sure what I can do about that.
> 
> @JaroCamphuijsen I was able to fix it by adding 'renv/sandbox' to the .gitignore
> 
> No idea why that causes the validation to choke, but nevermind...

I was just about to add that. Not sure why it suddenly started to worry about that line in the gitignore. Did something change since Tuesday when it would just pass all the tests?
Meh, anyhow, you solved it. Merge?

JaroCamphuijsen committed:
Remove line about agents

raar1 commented:
Yeah let's merge :)
